rgw: fix user rate limit is not enforced w/ global rate limit set

mkogan1 opened the pull request:
calling to `s->user->read_attrs(...);` similar to the way it was before commit https://github.com/ceph/ceph/pull/54333/commits/582970a69399d989370bb1ced6e42de724509622
which then is used by `get_attrs()` in `rate_limit(...)` at:
https://github.com/ceph/ceph/blob/4bd2fdc114e266aa67d09c9dda86e5fcc8da5d0c/src/rgw/rgw_process.cc#L115-L124

Fixes: https://tracker.ceph.com/issues/68346


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

cbodley commented:
thanks @mkogan1 (cc @dang)

auth is leaving `s->user` in a partially-initialized state because it only sets `RGWUserInfo` without attrs: https://github.com/ceph/ceph/blob/4bd2fdc/src/rgw/rgw_auth.cc#L526

can we fix the interface of `IdentityApplier::load_acct_info(const DoutPrefixProvider* dpp, RGWUserInfo& user_info)` so it can initialize both?

mkogan1 commented:
@cbodley, re:
> can we fix the interface of IdentityApplier::load_acct_info(const DoutPrefixProvider* dpp, RGWUserInfo& user_info) so it can initialize both?

pushed an update, please take a look

cbodley commented:
`s3::LocalEngine::authenticate()` is where we call `get_user_by_access_key()` to load the user info/attrs: https://github.com/ceph/ceph/blob/457aa80e/src/rgw/rgw_rest_s3.cc#L6342-L6346

that only passes the `RGWUserInfo` into the `IdentityApplier` through `apl_factory->create_apl_local()`: https://github.com/ceph/ceph/blob/457aa80e/src/rgw/rgw_rest_s3.cc#L6398-L6399

maybe `create_apl_local()` can take the whole `std::unique_ptr<rgw::sal::User>` there instead of just the `RGWUserInfo`? just trying to avoid needing a separate call to `read_attrs()` for every request

cbodley commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -6356,6 +6356,7 @@ rgw::auth::s3::LocalEngine::authenticate(
       throw -EPERM;
     }
   }*/
+  s->user->set_attrs(user->get_attrs());
```
hmm this call to `authenticate()` may end up failing, and a different auth engine could succeed with a different identity. this would use the wrong user's attrs for the request

mkogan1 replied:
@cbodley ack, thank you
pushed an update which moves the `s->user->set_attrs(...)` to a positions above a `return result_t::grant(...)`, at those positions checks have completed.

yet, two cases that not sure about are in:
https://github.com/ceph/ceph/blob/ee6523b618e7abe21cd2346f828f57cf150de5b9/src/rgw/rgw_rest_s3.cc#L6593
are `return result_t::grant(...)` from 
https://github.com/ceph/ceph/blob/ee6523b618e7abe21cd2346f828f57cf150de5b9/src/rgw/rgw_rest_s3.cc#L6691
and
https://github.com/ceph/ceph/blob/ee6523b618e7abe21cd2346f828f57cf150de5b9/src/rgw/rgw_rest_s3.cc#L6695
should also `s->user->set_attrs(...)` ?

cbodley commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -6494,6 +6494,7 @@ rgw::auth::s3::LocalEngine::authenticate(
 
   /* Ignore signature for HTTP OPTIONS */
   if (s->op_type == RGW_OP_OPTIONS_CORS) {
+    s->user->set_attrs(user->get_attrs());
     auto apl = apl_factory->create_apl_local(
         cct, s, user->get_info(), std::move(account), std::move(policies),
```
these auth engines get `const req_state*` intentionally to prevent them from making changes like this to the request. the fact that `s->user` is a pointer makes it possible to work around that restriction, but i'd prefer a design that doesn't

we're already passing `user->get_info()` in so that it can be copied into the request through `IdentityApplier::load_acct_info(const DoutPrefixProvider* dpp, RGWUserInfo& user_info)`. we call that [in a single place](https://github.com/ceph/ceph/blob/d908c9e/src/rgw/rgw_auth.cc#L524-L526) just after `authenticate()` succeeds:
```c++
      applier->load_acct_info(dpp, s->user->get_info());
```

i think the right thing to do is pass our initialized `user` (instead of just the `user->get_info()`) into `create_apl_local()`, and expand `load_acct_info()` so that it fully initializes `s->user`. for example,
```c++
virtual auto load_acct_info(const DoutPrefixProvider* dpp)
  -> std::unique_ptr<rgw::sal::User> = 0;
```
and
```c++
      s->user = applier->load_acct_info(dpp);
```

mkogan1 replied:
@cbodley hi, pushed an update - unfortunately the following did not work out:
> ```c++
> virtual auto load_acct_info(const DoutPrefixProvider* dpp)
>   -> std::unique_ptr<rgw::sal::User> = 0;
> ```  
> and
> ```c++
>       s->user = applier->load_acct_info(dpp);
> ```
because there are complication steaming from use cases in which the passed `RGWUserInfo& user_info` parameter is being acted upon so was not able to elide it.

https://github.com/ceph/ceph/blob/505f1a67a9238cdf4845a4f4b5bdbeeaf0f908c4/src/rgw/rgw_auth.cc#L634C1-L645C31

mkogan1 commented:
@cbodley PTAL at the latest push, addressing comment https://github.com/ceph/ceph/pull/60073#discussion_r1846935574

mkogan1 commented:
teuthology run results:
https://pulpito.ceph.com/mkogan-2024-12-03_08:51:40-rgw-wip-fix-rate-lim--i10-distro-default-smithi/

cbodley commented on src/rgw/rgw_auth.cc:
```diff
@@ -1121,6 +1123,22 @@ void rgw::auth::LocalApplier::write_ops_log_entry(rgw_log_entry& entry) const
 }
 
+rgw::auth::LocalApplier::LocalApplier(CephContext* const cct,
+                                      std::unique_ptr<rgw::sal::User>* user,
+                                      std::optional<RGWAccountInfo> account,
+                                      std::vector<IAM::Policy> policies,
+                                      std::string subuser,
+                                      const std::optional<uint32_t>& perm_mask,
+                                      const std::string access_key_id)
+  : user_info(user->get()->get_info()),
+    user(std::move(*user)),
```
since you're moving `*user`, the constructor should take `std::unique_ptr<rgw::sal::User>` by value instead of pointer. that requires the caller to pass `std::move(user)`, which makes it obvious at the call site that the `user` variable is being moved from

same goes for the `create_apl_local()` overloads

mkogan1 replied:
ack, done

mkogan1 commented on src/rgw/rgw_auth.cc:
```diff
@@ -523,7 +523,7 @@ rgw::auth::Strategy::apply(const DoutPrefixProvider *dpp, const rgw::auth::Strat
       /* Account used by a given RGWOp is decoupled from identity employed
        * in the authorization phase (RGWOp::verify_permissions). */
-      applier->load_acct_info(dpp, s->user->get_info());
+      s->user = applier->load_acct_info(dpp)->clone();
```
removed the `->clone()` seems that it is not necessary

mkogan1 commented on src/rgw/rgw_auth.cc:
```diff
@@ -294,9 +294,9 @@ static auto transform_old_authinfo(const RGWUserInfo& user,
-                        RGWUserInfo& user_info) const override {
+    auto load_acct_info(const DoutPrefixProvider* dpp) const -> std::unique_ptr<rgw::sal::User> override {
       // noop, this user info was passed in on construction
+      return nullptr;
```
ack, addressed

mkogan1 commented:
@cbodley hi, PTAL at this revision addressing the comments

teuthology run before the rebase:
https://pulpito.ceph.com/mkogan-2025-01-03_15:35:34-rgw-wip-fix-rate-lim--i10-07-distro-default-smithi/

mkogan1 commented:
jenkins test docs

mkogan1 commented:
jenkins test docs

mkogan1 commented:
jenkins render docs

cbodley commented:
> teuthology run before the rebase: https://pulpito.ceph.com/mkogan-2025-01-03_15:35:34-rgw-wip-fix-rate-lim--i10-07-distro-default-smithi/

looks good 👍 the two failures there are tracked in https://tracker.ceph.com/issues/69301 and https://tracker.ceph.com/issues/69441

mkogan1 commented:
teuthology run after rebase:
https://pulpito.ceph.com/mkogan-2025-01-06_14:09:11-rgw-wip-fix-rate-lim--i10-08-distro-default-smithi/
no anomalies (same test issue as mentioned in the comment above)
